src/engine/Join.cpp

```diff
@@ -729,3 +743,103 @@ void Join::doGallopInnerJoin(const TagType, const IdTableView<L_WIDTH>& l1,
     }
   }
 }
+
+// ______________________________________________________________________________
+template <int L_WIDTH, int R_WIDTH, int OUT_WIDTH>
+void Join::hashJoin(const IdTable& dynA, size_t jc1, const IdTable& dynB,
+                size_t jc2, IdTable* dynRes) {
+  const IdTableView<L_WIDTH> a = dynA.asStaticView<L_WIDTH>();
+  const IdTableView<R_WIDTH> b = dynB.asStaticView<R_WIDTH>();
+
+  LOG(DEBUG) << "Performing hashJoin between two tables.\n";
+  LOG(DEBUG) << "A: width = " << a.cols() << ", size = " << a.size() << "\n";
+  LOG(DEBUG) << "B: width = " << b.cols() << ", size = " << b.size() << "\n";
+
+  // Check trivial case.
+  if (a.size() == 0 || b.size() == 0) {
+    return;
+  }
+
+  IdTableStatic<OUT_WIDTH> result = dynRes->moveToStatic<OUT_WIDTH>();
+
+  // Cannot just switch l1 and l2 around because the order of
+  // items in the result tuples is important.
+  if (a.size() / b.size() > GALLOP_THRESHOLD) {
+    doGallopInnerJoin(LeftLargerTag{}, a, jc1, b, jc2, &result);
+  } else if (b.size() / a.size() > GALLOP_THRESHOLD) {
+    doGallopInnerJoin(RightLargerTag{}, a, jc1, b, jc2, &result);
+  // Procceding with the actual hash join depended on which IdTableView
+  // is bigger.
+  } else if (a.size() / b.size() >= 1) {
+    // a is bigger, or the same size as b. b gets put into the hash table.
+    ad_utility::HashMap<Id, std::vector<std::decay_t<typename IdTableStatic<R_WIDTH>::const_row_type>>> map;
+    for ( size_t j = 0; j < b.size(); j++) {
+      map[b(j, jc2)].push_back(b[j]);
+    }
+
+    // Create cross product by going through a.
+    for (size_t i = 0; i < a.size(); i++) {
+      // Skip, if there is no matching entry in the join column.
+      if (map.find(a(i, jc1)) == map.end()) {
+        continue;
+      }
+
+      for (const auto& bRow: map[a(i, jc1)]) {
+        const size_t backIndex = result.size();
+        result.push_back();
+        for (size_t h = 0; h < a.cols(); h++) {
+          result(backIndex, h) = a(i, h);
+        }
+
+        // Copy bs columns before the join column
+        for (size_t h = 0; h < jc2; h++) {
+          result(backIndex, h + a.cols()) = bRow[h];
+        }
+
+        // Copy bs columns after the join column
+        for (size_t h = jc2 + 1; h < b.cols(); h++) {
+          result(backIndex, h + a.cols() - 1) = bRow[h];
+        }
+      }
```

Comment:
Wouldn't it be wiser, to make `addCombinedRowFromResult` a private templated member function of the class, so that it later can be re-used for the normal `join` function? To also make that `join` function cleaner and all that.

Reply:
That is great thinking, 
you should definitely do this to further reduce code duplication
(The code quality of the existing joins is also questionable, so if you improve them along the way we are very thankful).

Reply:
Done. See commit 9e400bb.